cpp/src/reductions/scan/scan_inclusive.cu

```diff
@@ -198,22 +196,210 @@ struct scan_dispatcher {
+          element_equality_comparator<nested_nulls> element_comparator{col, col, true};
+          return type_dispatcher(col.type(), element_comparator, row_index, row_index - 1);
+        });
+    });
+}
+
+template <bool has_nulls>
+void generate_dense_rank_comparisons(column_view const& order_by,
+                                     mutable_column_view out,
+                                     rmm::cuda_stream_view stream)
+{
+  auto const d_order_by = column_device_view::create(order_by, stream);
+  element_equality_comparator<has_nulls> element_comparator(*d_order_by, *d_order_by, true);
+  thrust::tabulate(rmm::exec_policy(stream),
+                   out.begin<size_type>(),
+                   out.end<size_type>(),
+                   [type = d_order_by->type(), element_comparator] __device__(size_type row_index) {
+                     return (row_index == 0 ||
+                             !type_dispatcher(type, element_comparator, row_index, row_index - 1));
+                   });
+}
+
 }  // namespace
 
+std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+                                            rmm::cuda_stream_view stream,
+                                            rmm::mr::device_memory_resource* mr)
+{
+  auto ranks = make_fixed_width_column(
+    data_type{type_to_id<size_type>()}, order_by.size(), mask_state::ALL_VALID, stream, mr);
+  auto mutable_ranks = ranks->mutable_view();
+  if (order_by.type().id() == type_id::LIST) {
+    CUDF_FAIL("List types not supported");
+  } else if (order_by.type().id() == type_id::STRUCT) {
+    bool nested_nulls = struct_has_nested_nulls(order_by);
+    bool is_nested    = has_nested_children(order_by);
+
+    if (is_nested) {
+      CUDF_FAIL("Nested struct and list types not supported");
+    } else if (nested_nulls) {
+      generate_rank_struct_comparisons<true>(order_by, mutable_ranks, stream);
+    } else {
+      generate_rank_struct_comparisons<false>(order_by, mutable_ranks, stream);
+    }
+  } else if (order_by.has_nulls()) {
+    generate_rank_comparisons<true>(order_by, mutable_ranks, stream);
+  } else {
+    generate_rank_comparisons<false>(order_by, mutable_ranks, stream);
+  }
+
+  thrust::inclusive_scan(rmm::exec_policy(stream),
+                         mutable_ranks.begin<size_type>(),
+                         mutable_ranks.end<size_type>(),
+                         mutable_ranks.begin<size_type>(),
+                         DeviceMax{});
+  return ranks;
+}
+
+std::unique_ptr<column> inclusive_dense_rank_scan(column_view const& order_by,
+                                                  rmm::cuda_stream_view stream,
+                                                  rmm::mr::device_memory_resource* mr)
+{
+  auto ranks = make_fixed_width_column(
+    data_type{type_to_id<size_type>()}, order_by.size(), mask_state::ALL_VALID, stream, mr);
+  auto mutable_ranks = ranks->mutable_view();
```

Comment:
Instead of using an output parameter, it would be nice if the `ranks` column were returned from `generate_dense_rank_comparisons`.